cpp/src/arrow/compute/exec/key_hash.cc

```diff
@@ -186,11 +186,12 @@ void Hashing::hash_varlen_helper(uint32_t length, const uint8_t* key, uint32_t*
 
   int tail = length % 16;
   if (tail) {
-    uint64_t last_stripe[2];
-    const uint64_t* last_stripe_base =
+    uint32_t lanes[4];
+    uint64_t* const last_stripe = reinterpret_cast<uint64_t*>(lanes);
```

Comment:
This is turning fine code into obfuscated code for no good reason. Just zero-initialize `last_strip` if you want to silence this warning?

Reply:
I would defer to @michalursa here, but my thought was this was performance sensitive enough where we'd want to avoid the unnecessary initialization cost.

Reply:
The best way to find out is to run benchmarks.

Reply:
I tried getting it working with zero initializing `last_stripe` but it looks like this hits a compiler bug in GCC when using `-O3` and gets us to some undefined behavior which can lead to incorrect results: https://godbolt.org/z/4e9rbz8E4

Initializing the storage using `uint32_t` and then using `reinterpret_cast` to cast to a `uint64_t*` doesn't seem to hit the bug, but not sure how comfortable we should be using this generally.

Reply:
Well... More radical, though, is to question the existence of this function. We have another string hash function here:
https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/hashing.h#L130

Perhaps try to use it and see if performance differs?

Reply:
Tried a different approach in using a `memcpy` which the compiler _should_ optimize out. Will run the benchmarks and see if there's any impact.